llvm/lib/Target/AIE/AIEAddressSpaceFlattening.cpp

```diff
@@ -0,0 +1,124 @@
+// with AIE backend passes. Since we retype the pointer to P0, any
+// G_ADDRSPACE_CAST will have same i/p & o/p operand type which is not a valid
+// instruction. Thus we to replace them with COPY.
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseSubtarget.h"
+#include "llvm/CodeGen/GlobalISel/CSEInfo.h"
+#include "llvm/CodeGen/GlobalISel/CSEMIRBuilder.h"
+#include "llvm/CodeGen/GlobalISel/GenericMachineInstrs.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+
+#define DEBUG_TYPE "aie-addrspace-flattening"
+
+using namespace llvm;
+
+static const char AIE_ADDRSPACE_FLATTENING[] =
+    "AIE Address Space Flattening Pass";
+
+namespace {
+
+class AIEAddressSpaceFlattening : public MachineFunctionPass {
+public:
+  static char ID;
+  AIEAddressSpaceFlattening() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    TargetPassConfig &TPC = getAnalysis<TargetPassConfig>();
+    // Enable CSE to set up GISelObserverWrapper object
+    GISelCSEAnalysisWrapper &Wrapper =
+        getAnalysis<GISelCSEAnalysisWrapperPass>().getCSEWrapper();
+    auto *CSEInfo = &Wrapper.get(TPC.getCSEConfig());
+    std::unique_ptr<MachineIRBuilder> Builder =
+        CSEInfo ? std::make_unique<CSEMIRBuilder>()
+                : std::make_unique<MachineIRBuilder>();
+    Builder->setMF(MF);
+    MachineIRBuilder &MIB = *Builder;
+    // Set Observer to inform change of register type or change in op-code
+    GISelObserverWrapper Observer;
+    if (CSEInfo) {
+      Observer.addObserver(CSEInfo);
+      MIB.setChangeObserver(Observer);
+    }
+
+    bool Changed = false;
+    for (MachineBasicBlock &MBB : MF) {
+      for (MachineInstr &MI : MBB) {
+        Changed = processMI(MI, MRI, MIB, Observer);
+      }
+    }
+    return Changed;
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<MachineModuleInfoWrapperPass>();
+    AU.addRequired<GISelCSEAnalysisWrapperPass>();
+    AU.addRequired<TargetPassConfig>();
+    AU.setPreservesAll();
+  }
+
+  StringRef getPassName() const override { return AIE_ADDRSPACE_FLATTENING; }
+
+private:
+  bool processMI(MachineInstr &MI, MachineRegisterInfo &MRI,
+                 MachineIRBuilder &MIB, GISelObserverWrapper &Observer) {
+    bool Changed = false;
+    // Lower all pointers to pointers with default address space
+    for (llvm::MachineOperand &MO : MI.operands()) {
+      if (!MO.isReg())
+        continue;
+      const Register Reg = MO.getReg();
+      if (MRI.getType(Reg).isPointer() && Reg.isVirtual()) {
```

Comment:
I think you don't need  to check for `Reg.isVirtual()` because this is done in `MRI` anyway.